### PR TITLE
[mlir][python] python binding for the affine.store op

### PR DESCRIPTION
**Pull request by kaitingwang**
This PR creates the necessary files to support bindings for operations in the affine dialect.

This is the first of many PRs which will progressively introduce affine.load, affine.for, etc operations.  I would like to
acknowledge the work by Nelli's author @makslevental : https://github.com/makslevental/nelli/blob/main/nelli/mlir/affine/affine.py which jump-starts the work.

### CODE REVIEW

**Comment by makslevental** [comment]
Very cool. Thanks! Just one thing and then LGTM.

File: mlir/python/mlir/dialects/_affine_ops_ext.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+class AffineStoreOp:
+    """Specialization for the Affine store operation."""
+
+    def __init__(
+        self,
+        value: Value,
+        memref: Union[Operation, OpView, Value],
+        map,
+        *,
+        map_operands=[],
```
Comment (makslevental) [inline mlir/python/mlir/dialects/_affine_ops_ext.py:24]: This is one of those [python gotchas](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) that actually exists in other places in our code base. Can you make this `=None` and then add something like `map = map if map is not None else []` (not necessarily that concise - typing on my phone...).

**Review by makslevental — CHANGES_REQUESTED** [review]

File: mlir/python/mlir/dialects/AffineOps.td [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#ifndef PYTHON_BINDINGS_AFFINE_OPS
+#define PYTHON_BINDINGS_AFFINE_OPS
+
+include "mlir/Dialect/Affine/IR/AffineOps.td"
+
+#endif // PYTHON_BINDINGS_AFFINE_OPS
```
Comment (ftynse) [inline mlir/python/mlir/dialects/AffineOps.td:6]: Trailing newline please.

File: mlir/python/mlir/dialects/_affine_ops_ext.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+from typing import Optional, Sequence, Union
+
+
+class AffineStoreOp:
+    """Specialization for the Affine store operation."""
+
+    def __init__(
+        self,
+        value: Value,
```
Comment (ftynse) [inline mlir/python/mlir/dialects/_affine_ops_ext.py:20]: This should also be a union so we can take an operation and "unpack" it to get the result (if single). Similarly to the other argument. Note that the implementation handles this correctly.

File: mlir/python/mlir/dialects/_affine_ops_ext.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        memref: Union[Operation, OpView, Value],
+        map,
```
Comment (ftynse) [inline mlir/python/mlir/dialects/_affine_ops_ext.py:22]: Please add a type annotation.

File: mlir/python/mlir/dialects/affine.py [hunk 5]
```diff
@@ -0,0 +1 @@
+from ._affine_ops_gen import *
```
Comment (ftynse) [inline mlir/python/mlir/dialects/affine.py:1]: Newline please.

File: mlir/python/mlir/dialects/_affine_ops_ext.py (hunk 1 above)
Comment (kaitingwang) [reply mlir/python/mlir/dialects/_affine_ops_ext.py:24]: Thanks!  In fact that map = map ...  is concise (error free).   Updated accordingly.

File: mlir/python/mlir/dialects/AffineOps.td (hunk 2 above)
Comment (kaitingwang) [reply mlir/python/mlir/dialects/AffineOps.td]: Fixed.

File: mlir/python/mlir/dialects/_affine_ops_ext.py (hunk 3 above)
Comment (kaitingwang) [reply mlir/python/mlir/dialects/_affine_ops_ext.py]: Fixed to be the same as the other argument (i.e. Union[Operation, OpView, Value]).  Thank you!

File: mlir/python/mlir/dialects/affine.py (hunk 5 above)
Comment (kaitingwang) [reply mlir/python/mlir/dialects/affine.py]: Fixed and took the opportunity to add the copyright template into it as well.  Thanks!

File: mlir/python/mlir/dialects/_affine_ops_ext.py (hunk 4 above)
Comment (kaitingwang) [reply mlir/python/mlir/dialects/_affine_ops_ext.py]: Fixed.

File: mlir/python/mlir/dialects/_affine_ops_ext.py (hunk 1 above)
Comment (makslevental) [reply mlir/python/mlir/dialects/_affine_ops_ext.py:24]: same for `map_operands` 🙂

Comment (kaitingwang) [reply mlir/python/mlir/dialects/_affine_ops_ext.py]: Fixed. :)  I still kept 2 commits so the fix for map_operands is squashed into previous review commit.

**Review by ftynse — APPROVED** [review]

**Review by makslevental — APPROVED** [review]

**Comment by makslevental** [comment]
@kaitingwang let me know if you need me to merge.

**Comment by kaitingwang** [comment]
@makslevental @ftynse Appreciated the reviews.
@makslevental Thanks for the offer.  A single button click so I merged it myself just now.  :) Will likely send more PRs down your way in the days to come!

**Pull request merged**